components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+    andor_polarity_operand: u8,
+    modulo: <u32 as AsULE>::ULE,
+    range_list: [RangeOrValueULE],
+}
+
+impl RelationULE {
+    pub fn as_relation(&self) -> Relation {
+        // safe to call because we are operating on a valid RelationULE
+        let (and_or, polarity, operand) =
+            unsafe { Self::decode_andor_polarity_operand(self.andor_polarity_operand) };
+        Relation {
+            and_or,
+            polarity,
+            operand,
+            modulo: u32::from_unaligned(self.modulo),
+            range_list: ZeroVec::borrowed_from_slice(&self.range_list),
+        }
+    }
+
+    fn encode_andor_polarity_operand(and_or: AndOr, polarity: Polarity, operand: Operand) -> u8 {
+        // XXX: Ensure and_or is one bit, polarity is one bit, and operand is max 6 bits
+        (((and_or == AndOr::And) as u8) << 7)
+            + (((polarity == Polarity::Positive) as u8) << 6)
+            + operand.idx()
+    }
+
+    // Must be called on an `encoded` that is obtained from `encode_andor_polarity_operand`, i.e.
+    // the field on a valid RelationULE
+    unsafe fn decode_andor_polarity_operand(encoded: u8) -> (AndOr, Polarity, Operand) {
+        let and_or = if encoded & 0b1000_0000 != 0 {
+            AndOr::And
+        } else {
+            AndOr::Or
+        };
+
+        let polarity = if encoded & 0b0100_0000 != 0 {
+            Polarity::Positive
+        } else {
+            Polarity::Negative
+        };
+
+        let operand = Operand::from_idx(encoded & 0b0011_1111);
+        (and_or, polarity, operand)
+    }
+}
+
+unsafe impl VarULE for RelationULE {
+    type Error = &'static str;
+
+    #[inline]
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
+        let ptr = bytes.as_ptr();
+        let len = bytes.len();
+        // subtract length of andor_polarity_operand and modulo and then convert between a slice of bytes and PlainOldULE<8>
+        let len_new = (len - 5) / 8;
+        // it's hard constructing custom DSTs, we fake a pointer/length construction
+        // eventually we can use the Pointer::Metadata APIs when they stabilize
+        let fake_slice = core::ptr::slice_from_raw_parts(ptr as *const RangeOrValueULE, len_new);
+        let ret = &*(fake_slice as *const Self);
+        debug_assert_eq!(core::mem::size_of_val(ret), core::mem::size_of_val(bytes));
+        ret
+    }
+
+    #[inline]
+    fn as_byte_slice(&self) -> &[u8] {
+        unsafe {
+            core::slice::from_raw_parts(
+                self as *const Self as *const u8,
+                core::mem::size_of_val(self),
+            )
+        }
+    }
+
+    //     what it should do is attempt to parse the first 4 bytes as a u32::ULE (POU<4>), and the remaining bytes as a ZV<RangeOrValueULE>
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
+        <PlainOldULE<4> as ULE>::validate_byte_slice(&bytes[1..5]).map_err(|_| "foo")?;
```

Comment:
Nit: `<PlainOldULE<4> as ULE>::validate_byte_slice` is infallible. You can either mark the error case as `unreachable!()` or just delete this line.